### PR TITLE
Using Polly library instead of TOPAZ (as it's outdated), for SQL with…

### PR DESCRIPTION
**Pull request by luisgmsft**
… ADO.NET

### CODE REVIEW

**Comment by bennage** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/best-practices/retry-service-specific.md](https://github.com/luisgmsft/architecture-center/blob/retry-specifics-guidance/sqladonet-polly/docs/best-practices/retry-service-specific.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/azure/architecture/best-practices/retry-service-specific?branch=pr-en-us-80) |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2017%5C5%5C4%5C0659c0de-90b3-6253-8249-f3f6e8fba551%5CPullRequest%5C201705040008265390-80%5Cworkflow_report.html).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

File: docs/best-practices/retry-service-specific.md [hunk 1]
```diff
@@ -423,33 +384,77 @@ However, in the current version of the Transient Fault Handling Application Bloc
...
+        await Task.Delay(TimeSpan.FromSeconds(0.5), cancellationToken);
+    }
+
+    Console.WriteLine("");
+    Console.WriteLine($"Total requests made                 : {i}" + i);
+    Console.WriteLine($"Requests which eventually succeeded : {eventualSuccesses}");
+    Console.WriteLine($"Retries made to help achieve success: {retries}");
+    Console.WriteLine($"Requests which eventually failed    : {eventualFailures}");
 }
 ```
```
Comment (reisenberger) [inline docs/best-practices/retry-service-specific.md]: Thank you for mentioning Polly in this guidance. We are glad simply if [Polly](https://github.com/App-vNext/Polly) is useful to the wider .NET community. I am offering some comments as (disclosure) [lead developer on the Polly library](https://github.com/App-vNext/Polly/graphs/contributors).

The code example for Polly with ADO.NET could be simplified, per code suggestion below:

```csharp
public async static Task<SqlDataReader> ExecuteReaderWithRetryAsync(this SqlCommand command)
{
    GuardConnectionIsNotNull(command);

    var policy = Policy.Handle<Exception>().WaitAndRetryAsync(
        retryCount: 3, // Retry 3 times
        sleepDurationProvider: attempt => TimeSpan.FromMilliseconds(200 * Math.Pow(2, attempt - 1)), // Exponential backoff based on an initial 200ms delay.
        onRetry: (exception, attempt) => 
        {
            // Capture some info for logging/telemetry.  
            logger.LogWarn($"ExecuteReaderWithRetryAsync: Retry {attempt} due to {exception}.");
        });

    // Retry the following call according to the policy.
    await policy.ExecuteAsync<SqlDataReader>(async token =>
    {
        // This code is executed within the Policy 

        if (conn.State != System.Data.ConnectionState.Open) await conn.OpenAsync(token);
        return await command.ExecuteReaderAsync(System.Data.CommandBehavior.Default, token);

    }, cancellationToken);
}
```

Changes from the original are:

+ Remove material around `Console.ReadKey()`, `Console.Writeline()`, which was only relevant to the [Polly-Samples demo](https://github.com/App-vNext/Polly-Samples) you originally referenced, and is not relevant in production.
+ Replace with example logging.
+ Change wait between retries to an exponential backoff example with Polly (following the recommendation you have for exponential backoff with SQL, earlier in the article)
+ Change all SQL calls to async calls taking advantage of the passed-in `CancellationToken`

Hope that this is useful.

**Comment by bennage** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/best-practices/retry-service-specific.md](https://github.com/luisgmsft/architecture-center/blob/retry-specifics-guidance/sqladonet-polly/docs/best-practices/retry-service-specific.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/azure/architecture/best-practices/retry-service-specific?branch=pr-en-us-80) |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2017%5C5%5C8%5C0659c0de-90b3-6253-8249-f3f6e8fba551%5CPullRequest%5C201705081906294905-80%5Cworkflow_report.html).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Comment by bennage** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/best-practices/retry-service-specific.md](https://github.com/luisgmsft/architecture-center/blob/retry-specifics-guidance/sqladonet-polly/docs/best-practices/retry-service-specific.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/azure/architecture/best-practices/retry-service-specific?branch=pr-en-us-80) |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2017%5C5%5C12%5C0659c0de-90b3-6253-8249-f3f6e8fba551%5CPullRequest%5C201705122258569284-80%5Cworkflow_report.html).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Comment by bennage** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/best-practices/retry-service-specific.md](https://github.com/luisgmsft/architecture-center/blob/retry-specifics-guidance/sqladonet-polly/docs/best-practices/retry-service-specific.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/azure/architecture/best-practices/retry-service-specific?branch=pr-en-us-80) |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2017%5C5%5C12%5C0659c0de-90b3-6253-8249-f3f6e8fba551%5CPullRequest%5C201705122303038650-80%5Cworkflow_report.html).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Pull request merged**